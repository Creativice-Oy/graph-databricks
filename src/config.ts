import {
  IntegrationExecutionContext,
  IntegrationValidationError,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  databricksSubdomain: {
    type: 'string',
  },
  databricksAccessToken: {
    type: 'string',
    mask: true,
  },
};

export interface IntegrationConfig extends IntegrationInstanceConfig {
  databricksSubdomain: string;
  databricksAccessToken: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;

  if (!config.databricksSubdomain || !config.databricksAccessToken) {
    throw new IntegrationValidationError(
      'Config requires all of {databricksSubdomain, databricksAccessToken}',
    );
  }

  const apiClient = createAPIClient(config);
  await apiClient.verifyAuthentication();
}
