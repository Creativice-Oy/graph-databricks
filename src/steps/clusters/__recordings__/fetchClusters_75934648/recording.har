{
  "log": {
    "_recordingName": "fetchClusters",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "bf48bfc2950b87846e3443d5d700288c",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "3840207185082406.6.gcp.databricks.com"
            }
          ],
          "headersSize": 320,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://3840207185082406.6.gcp.databricks.com/api/2.0/clusters/list"
        },
        "response": {
          "bodySize": 1977,
          "content": {
            "mimeType": "application/json",
            "size": 1977,
            "text": "{\"clusters\":[{\"cluster_id\":\"0201-183853-7c8r8974\",\"driver\":{\"private_ip\":\"127.0.0.1\"},\"spark_context_id\":8745868079977237257,\"cluster_name\":\"test-cluster\",\"spark_version\":\"9.1.x-scala2.12\",\"spark_conf\":{\"spark.master\":\"local[*, 4]\",\"spark.databricks.cluster.profile\":\"singleNode\"},\"node_type_id\":\"n1-standard-4\",\"driver_node_type_id\":\"n1-standard-4\",\"custom_tags\":{\"ResourceClass\":\"SingleNode\"},\"spark_env_vars\":{\"PYSPARK_PYTHON\":\"/databricks/python3/bin/python3\"},\"autotermination_minutes\":20,\"enable_elastic_disk\":false,\"disk_spec\":{},\"cluster_source\":\"UI\",\"enable_local_disk_encryption\":false,\"gcp_attributes\":{\"use_preemptible_executors\":false,\"availability\":\"ON_DEMAND_GCP\",\"zone_id\":\"HA\"},\"instance_source\":{\"node_type_id\":\"n1-standard-4\"},\"driver_instance_source\":{\"node_type_id\":\"n1-standard-4\"},\"state\":\"RUNNING\",\"start_time\":1643740733901,\"last_state_loss_time\":1643741512861,\"num_workers\":0,\"cluster_memory_mb\":15360,\"cluster_cores\":4.0,\"default_tags\":{\"Vendor\":\"Databricks\",\"Creator\":\"viragsf_at_gmail.com\",\"ClusterName\":\"test-cluster\",\"ClusterId\":\"0201-183853-7c8r8974\"},\"creator_user_name\":\"viragsf@gmail.com\",\"init_scripts_safe_mode\":false},{\"cluster_id\":\"0117-142331-yanks462\",\"driver\":{\"private_ip\":\"127.0.0.1\"},\"cluster_name\":\"test-gcp-cluster\",\"spark_version\":\"9.1.x-scala2.12\",\"node_type_id\":\"n1-standard-4\",\"driver_node_type_id\":\"n1-standard-4\",\"spark_env_vars\":{\"PYSPARK_PYTHON\":\"/databricks/python3/bin/python3\"},\"autotermination_minutes\":20,\"enable_elastic_disk\":false,\"disk_spec\":{},\"cluster_source\":\"UI\",\"enable_local_disk_encryption\":false,\"gcp_attributes\":{\"use_preemptible_executors\":false,\"availability\":\"ON_DEMAND_GCP\",\"zone_id\":\"HA\"},\"instance_source\":{\"node_type_id\":\"n1-standard-4\"},\"driver_instance_source\":{\"node_type_id\":\"n1-standard-4\"},\"state\":\"TERMINATED\",\"state_message\":\"Cluster terminated by CLOUD_PROVIDER_LAUNCH_FAILURE\",\"start_time\":1642429411156,\"terminated_time\":1642429429719,\"last_state_loss_time\":1642429429732,\"autoscale\":{\"min_workers\":1,\"max_workers\":2},\"default_tags\":{\"Vendor\":\"Databricks\",\"Creator\":\"austin.kelleher_at_jupiterone.dev\",\"ClusterName\":\"test-gcp-cluster\",\"ClusterId\":\"0117-142331-yanks462\"},\"creator_user_name\":\"austin.kelleher@jupiterone.dev\",\"termination_reason\":{\"code\":\"CLOUD_PROVIDER_LAUNCH_FAILURE\",\"parameters\":{\"gcp_error_message\":\"Compute Quota Exceeded for jupiterone-databricks-test in region us-west1: Quota: SSD_TOTAL_GB, used 0.0 and requested 1000.0 out of 500.0\"},\"type\":\"CLOUD_FAILURE\"},\"init_scripts_safe_mode\":false}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "databricks"
            },
            {
              "name": "date",
              "value": "Tue, 01 Feb 2022 19:02:56 GMT"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-databricks-org-id",
              "value": "3840207185082406"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 330,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-02-01T19:02:55.914Z",
        "time": 772,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 772
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
