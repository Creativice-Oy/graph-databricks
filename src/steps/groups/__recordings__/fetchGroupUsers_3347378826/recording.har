{
  "log": {
    "_recordingName": "fetchGroupUsers",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "19c964637ad1008e1e4a38fd004b743d",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "3840207185082406.6.gcp.databricks.com"
            }
          ],
          "headersSize": 318,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://3840207185082406.6.gcp.databricks.com/api/2.0/groups/list"
        },
        "response": {
          "bodySize": 143,
          "content": {
            "mimeType": "application/json",
            "size": 143,
            "text": "{\"group_names\":[\"admins\",\"test-group-1\",\"test-group-2\"]}\n"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "databricks"
            },
            {
              "name": "x-databricks-reason-phrase",
              "value": "OK"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "x-databricks-server-name",
              "value": "cons-webapp-0"
            },
            {
              "name": "x-databricks-route-to",
              "value": "https://cons-webapp-0.cons-webapp-r.production.svc.cluster.local"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, User-Agent"
            },
            {
              "name": "date",
              "value": "Tue, 01 Feb 2022 18:27:33 GMT"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-databricks-org-id",
              "value": "3840207185082406"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 566,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-02-01T18:27:32.838Z",
        "time": 754,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 754
        }
      },
      {
        "_id": "2be559309a689f84ba4f3055c1624c66",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "3840207185082406.6.gcp.databricks.com"
            }
          ],
          "headersSize": 344,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "group_name",
              "value": "admins"
            }
          ],
          "url": "https://3840207185082406.6.gcp.databricks.com/api/2.0/groups/list-members?group_name=admins"
        },
        "response": {
          "bodySize": 221,
          "content": {
            "mimeType": "application/json",
            "size": 221,
            "text": "{\"members\":[{\"user_name\":\"viragsf@gmail.com\"},{\"user_name\":\"austin.kelleher@jupiterone.dev\"}]}\n"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "databricks"
            },
            {
              "name": "x-databricks-reason-phrase",
              "value": "OK"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, User-Agent"
            },
            {
              "name": "x-databricks-server-name",
              "value": "cons-webapp-0"
            },
            {
              "name": "x-databricks-route-to",
              "value": "https://cons-webapp-0.cons-webapp-r.production.svc.cluster.local"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "date",
              "value": "Tue, 01 Feb 2022 18:27:34 GMT"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-databricks-org-id",
              "value": "3840207185082406"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 566,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-02-01T18:27:33.604Z",
        "time": 834,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 834
        }
      },
      {
        "_id": "fc777b7c1600852ad7edc74cf4137122",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "3840207185082406.6.gcp.databricks.com"
            }
          ],
          "headersSize": 350,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "group_name",
              "value": "test-group-1"
            }
          ],
          "url": "https://3840207185082406.6.gcp.databricks.com/api/2.0/groups/list-members?group_name=test-group-1"
        },
        "response": {
          "bodySize": 3,
          "content": {
            "mimeType": "application/json",
            "size": 3,
            "text": "{}\n"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "databricks"
            },
            {
              "name": "x-databricks-reason-phrase",
              "value": "OK"
            },
            {
              "name": "x-databricks-server-name",
              "value": "cons-webapp-0"
            },
            {
              "name": "content-length",
              "value": "3"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "date",
              "value": "Tue, 01 Feb 2022 18:27:34 GMT"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-databricks-route-to",
              "value": "https://cons-webapp-0.cons-webapp-r.production.svc.cluster.local"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, User-Agent"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-databricks-org-id",
              "value": "3840207185082406"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 533,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-02-01T18:27:34.487Z",
        "time": 705,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 705
        }
      },
      {
        "_id": "5a335c2b86793642fe3d0cc38d3b9bf0",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "3840207185082406.6.gcp.databricks.com"
            }
          ],
          "headersSize": 350,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "group_name",
              "value": "test-group-2"
            }
          ],
          "url": "https://3840207185082406.6.gcp.databricks.com/api/2.0/groups/list-members?group_name=test-group-2"
        },
        "response": {
          "bodySize": 3,
          "content": {
            "mimeType": "application/json",
            "size": 3,
            "text": "{}\n"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "databricks"
            },
            {
              "name": "x-databricks-reason-phrase",
              "value": "OK"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "date",
              "value": "Tue, 01 Feb 2022 18:27:35 GMT"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, User-Agent"
            },
            {
              "name": "x-databricks-server-name",
              "value": "cons-webapp-0"
            },
            {
              "name": "content-length",
              "value": "3"
            },
            {
              "name": "x-databricks-route-to",
              "value": "https://cons-webapp-0.cons-webapp-r.production.svc.cluster.local"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-databricks-org-id",
              "value": "3840207185082406"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 533,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-02-01T18:27:35.196Z",
        "time": 857,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 857
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
