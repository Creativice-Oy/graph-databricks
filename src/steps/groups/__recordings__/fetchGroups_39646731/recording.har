{
  "log": {
    "_recordingName": "fetchGroups",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "d31c258b4eecb943136691f6a1410ecb",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "dbc-c50dbe80-ed72.cloud.databricks.com"
            }
          ],
          "headersSize": 320,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://dbc-c50dbe80-ed72.cloud.databricks.com/api/2.0/groups/list"
        },
        "response": {
          "bodySize": 27,
          "content": {
            "mimeType": "application/json",
            "size": 27,
            "text": "{\"group_names\":[\"admins\"]}\n"
          },
          "cookies": [],
          "headers": [
            {
              "name": "server",
              "value": "databricks"
            },
            {
              "name": "x-databricks-reason-phrase",
              "value": "OK"
            },
            {
              "name": "date",
              "value": "Tue, 01 Mar 2022 17:20:53 GMT"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "x-databricks-server-name",
              "value": "cons-webapp-1"
            },
            {
              "name": "content-length",
              "value": "27"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, User-Agent"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-databricks-route-to",
              "value": "https://cons-webapp-1.cons-webapp-r.production.svc.cluster.local"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "content-type",
              "value": "application/json"
            },
            {
              "name": "x-databricks-org-id",
              "value": "1393914232801712"
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 534,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-03-01T17:20:51.877Z",
        "time": 1571,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 1571
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
