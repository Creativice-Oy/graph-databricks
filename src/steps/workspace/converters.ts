import {
  createIntegrationEntity,
  Entity,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';

export function createWorkspaceEntity(subdomain: string): Entity {
  return createIntegrationEntity({
    entityData: {
      source: {
        id: subdomain,
      },
      assign: {
        _key: `databricks_workspace:${subdomain}`,
        _type: Entities.WORKSPACE._type,
        _class: Entities.WORKSPACE._class,
        name: subdomain,
      },
    },
  });
}
