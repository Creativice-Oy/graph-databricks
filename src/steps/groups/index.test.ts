import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { fetchGroups, fetchGroupUsers } from '.';
import { integrationConfig } from '../../../test/config';
import { setupDatabricksRecording } from '../../../test/recording';
import { IntegrationConfig } from '../../config';
import { Relationships } from '../constants';
import { fetchWorkspaceDetails } from '../workspace';

describe('#fetchGroups', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupDatabricksRecording({
      directory: __dirname,
      name: 'fetchGroups',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchWorkspaceDetails(context);
    await fetchGroups(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      collectedEntities: context.jobState.collectedEntities,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const groups = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('databricks_group'),
    );

    expect(groups.length).toBeGreaterThan(0);
    expect(groups).toMatchGraphObjectSchema({
      _class: ['UserGroup'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'databricks_group' },
          name: { type: 'string' },
          displayName: { type: 'string' },
        },
      },
    });

    const workspaceHasGroups = context.jobState.collectedRelationships.filter(
      (e) => e._type === Relationships.WORKSPACE_HAS_GROUP._type,
    );

    expect(workspaceHasGroups.length).toBeGreaterThan(0);
    expect(workspaceHasGroups).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: 'databricks_workspace_has_group',
          },
        },
      },
    });
  });
});

describe('#fetchGroupUsers', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupDatabricksRecording({
      directory: __dirname,
      name: 'fetchGroupUsers',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchWorkspaceDetails(context);
    await fetchGroups(context);
    await fetchGroupUsers(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      collectedEntities: context.jobState.collectedEntities,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const users = context.jobState.collectedEntities.filter((e) =>
      e._type.includes('databricks_user'),
    );

    expect(users.length).toBeGreaterThan(0);
    expect(users).toMatchGraphObjectSchema({
      _class: ['User'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'databricks_user' },
          name: { type: 'string' },
          username: { type: 'string' },
        },
      },
    });

    const groupHasUsers = context.jobState.collectedRelationships.filter(
      (e) => e._type === Relationships.GROUP_HAS_USER._type,
    );

    expect(groupHasUsers.length).toBeGreaterThan(0);
    expect(groupHasUsers).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: 'databricks_group_has_user',
          },
        },
      },
    });
  });
});
