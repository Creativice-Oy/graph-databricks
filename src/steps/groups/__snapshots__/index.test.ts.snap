// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchGroupUsers should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "databricks_workspace:dbc-c50dbe80-ed72.cloud.databricks.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "dbc-c50dbe80-ed72.cloud.databricks.com",
          },
        },
      ],
      "_type": "databricks_workspace",
      "createdOn": undefined,
      "displayName": "dbc-c50dbe80-ed72.cloud.databricks.com",
      "id": "dbc-c50dbe80-ed72.cloud.databricks.com",
      "name": "dbc-c50dbe80-ed72.cloud.databricks.com",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "databricks_group:admins",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "name": "admins",
          },
        },
      ],
      "_type": "databricks_group",
      "createdOn": undefined,
      "displayName": "admins",
      "name": "admins",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "databricks_user:stefan.virag@contractor.jupiterone.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "username": "stefan.virag@contractor.jupiterone.com",
          },
        },
      ],
      "_type": "databricks_user",
      "createdOn": undefined,
      "displayName": "stefan.virag@contractor.jupiterone.com",
      "name": "stefan.virag@contractor.jupiterone.com",
      "username": "stefan.virag@contractor.jupiterone.com",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "databricks_user:austin.kelleher@jupiterone.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "username": "austin.kelleher@jupiterone.com",
          },
        },
      ],
      "_type": "databricks_user",
      "createdOn": undefined,
      "displayName": "austin.kelleher@jupiterone.com",
      "name": "austin.kelleher@jupiterone.com",
      "username": "austin.kelleher@jupiterone.com",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "databricks_user:sam.andrus@jupiterone.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "username": "sam.andrus@jupiterone.com",
          },
        },
      ],
      "_type": "databricks_user",
      "createdOn": undefined,
      "displayName": "sam.andrus@jupiterone.com",
      "name": "sam.andrus@jupiterone.com",
      "username": "sam.andrus@jupiterone.com",
    },
  ],
  "encounteredTypes": Array [
    "databricks_workspace",
    "databricks_group",
    "databricks_workspace_has_group",
    "databricks_user",
    "databricks_group_has_user",
  ],
  "numCollectedEntities": 5,
}
`;

exports[`#fetchGroups should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "databricks_workspace:dbc-c50dbe80-ed72.cloud.databricks.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "dbc-c50dbe80-ed72.cloud.databricks.com",
          },
        },
      ],
      "_type": "databricks_workspace",
      "createdOn": undefined,
      "displayName": "dbc-c50dbe80-ed72.cloud.databricks.com",
      "id": "dbc-c50dbe80-ed72.cloud.databricks.com",
      "name": "dbc-c50dbe80-ed72.cloud.databricks.com",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "databricks_group:admins",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "name": "admins",
          },
        },
      ],
      "_type": "databricks_group",
      "createdOn": undefined,
      "displayName": "admins",
      "name": "admins",
    },
  ],
  "encounteredTypes": Array [
    "databricks_workspace",
    "databricks_group",
    "databricks_workspace_has_group",
  ],
  "numCollectedEntities": 2,
}
`;
